monix-eval/shared/src/main/scala/monix/eval/internal/TaskCreate.scala

```diff
@@ -184,44 +166,85 @@ private[eval] object TaskCreate {
-    new Callback[Throwable, A] {
-      override def onSuccess(value: A): Unit =
-        try {
-          ctx.scheduler.execute(new Runnable {
-            def run(): Unit = {
-              ctx.frameRef.reset()
-              cb.onSuccess(value)
-            }
-          })
-        } catch {
-          case e: RejectedExecutionException =>
-            Callback
-              .trampolined(cb)(ctx.scheduler)
-              .onError(e)
-        }
-
-      override def onError(e: Throwable): Unit =
-        try {
-          ctx.scheduler.execute(new Runnable {
-            def run(): Unit = {
-              ctx.frameRef.reset()
-              cb.onError(e)
-            }
-          })
-        } catch {
-          case e: RejectedExecutionException =>
-            Callback
-              .trampolined(cb)(ctx.scheduler)
-              .onError(e)
-        }
+  private[internal] def protectedCallback[A](
+    ctx: Context,
+    shouldPop: Boolean,
+    cb: Callback[Throwable, A]): Callback[Throwable, A] =
+    new CallbackForCreate[A](ctx, shouldPop, cb)
+
+  private final class CallbackForCreate[A](ctx: Context, threadId: Long, shouldPop: Boolean, cb: Callback[Throwable, A])
+    extends Callback[Throwable, A] with TrampolinedRunnable {
+
+    private[this] val state = AtomicInt(0)
+    private[this] var value: A = _
+    private[this] var error: Throwable = _
+    private[this] var isSameThread = false
+
+    def this(ctx: Context, shouldPop: Boolean, cb: Callback[Throwable, A]) =
+      this(ctx, Platform.currentThreadId(), shouldPop, cb)
+
+    override def onSuccess(value: A): Unit = {
+      if (state.compareAndSet(0, 1)) {
+        this.value = value
+        startExecution()
+      } else {
+        throw new IllegalStateException("Callback.onSuccess signaled multiple times")
+      }
+    }
+
+    override def onError(e: Throwable): Unit = {
+      if (state.compareAndSet(0, 2)) {
+        this.error = e
+        startExecution()
+      } else {
+        throw new IllegalStateException("Callback.onSuccess onError multiple times", e)
+      }
+    }
+
+    private def startExecution(): Unit = {
+      // Optimization — if the callback was called on the same thread
+      // where it was created, then we are not going to fork
+      isSameThread = Platform.currentThreadId() == threadId
+      try {
+        ctx.scheduler.execute(
+          if (isSameThread)
+            this
+          else
+            StartAsyncBatchRunnable(this, ctx.scheduler)
+        )
+      } catch {
+        case e: RejectedExecutionException =>
+          if (error != null) {
+            val e2 = error
+            error = null
+            ctx.scheduler.reportFailure(e2)
+          }
+          state.set(2)
+          this.value = null.asInstanceOf[A]
+          this.error = e
+          isSameThread = true
+          TrampolineExecutionContext.immediate.execute(this)
+      }
+    }
+
+    override def run(): Unit = {
+      if (!isSameThread) {
```

Comment:
I'm not sure I understand how it works. I will appreciate explanation

We create this `Callback` / `Runnable` in `Task.create`. At some point we run it, `isSameThread` is `false` so we reset the frameRef and then go to `startExecution()`, setting `isSameThread` variable to `true` or `false`. So I assume the instance of this callback is reused somewhere? Is it for each time given `Task.create` is used? Why do we reset it at least once?

Reply:
This is my understanding:

`StartAsyncBatchRunnable` forces an async boundary if we're on the other thread
Resetting frameRef ensures that internal batching mechanism isn't deoptimized by thinking there has been a lot of runs without async boundaries. This is only done if we're inserting the async boundary.
We only call methods once. `startExecution` is called first, upon completion, to set the `isSameThread`. `run` is called later via scheduling.

Reply:
Thanks, I misread which `onSuccess` / `onError` is called in `run()`, makes sense now

Reply:
An additional detail ... the callback inherits from `TrampolinedRunnable`. If executed as is (`scheduler.execute(this)`), then it gets executed via the trampoline on the current thread.

On the other hand by wrapping it in a `StartAsyncBatchRunnable`, we ensure that a fork happens.